hashintel/hash
# Automate Type URI generation

## 🌟 What is the purpose of this PR?

When Types are made through the workspace, we don't want users having to manually choose appropriate (and valid) URIs.
This PR abstracts that way, automating the process of the URIs being generated. It makes a large number of drivebys as well, to documentation, testing strategy, etc. (These were somewhat necessary in development, apologies they weren't separated out better).

There were also other concerns like making sure URIs are valid and that the workspace is capable of serving them.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202848989198299/f) _(internal)_

## 🔍 What does this change?

- Adds the `DomainValidator` struct in the Graph which uses a configured Regex to validate type URIs.
- Changes the signature of all `update*` methods on the Graph API to take a schema without an $id, and now an additional parameter `typeToUpdate`
  - Changes the implementation of the entry point of those methods to increment the `typeToUpdate` URI, and to inject it into the schema before reconstructing a full type
- Adds the `VAR_UPDATE_*` json schemas to use in the Graph API methods

- Completely reworks how seed data is created/ensured on the Workspace. A possibly too-magic implementation of async function initializers is used to handle initialization logic and dependencies.
- Modifies all resolvers to handle the new signatures
- `create*` resolvers for communicating with the frontend now take `Omit<type, "$id">` and generate a type at that time using the given accountId (which requires finding the appropriate shortname)

- Rewrites all relevant tests to create types on appropriate domains

## 📜 Does this require a change to the docs?

- The generated documentation has been adjusted. Some additional improvements have been added to the READMEs.
- We almost certainly need to improve docs around running tests, ensuring a clean DB, etc.

## ⚠️ Known issues
- There is a race condition inside the Graph where we check for an existing URI. This means we have to run integration tests sequentially
- Many more, see comment threads @Alfred-Mountfield has opened. Ensure these are resolved before merge.

## 🐾 Next steps

- Also many, see comment threads and relevant TODOs in the codebase (asana tasks have been included)

## 🛡 What tests cover this?

- Rust unit tests
- Rust integration tests
- HTTP tests for the Graph REST API
- Workspace integration tests for:
  -  "knowledge"
  - "ontology"

## ❓ How to test this?

1.  Checkout the branch / view the deployment
2. Try running the Graph tests according to the readme.
3. Ensure you take down the docker deployment and clear all volumes.
4. Run `external-services up --build` in the repo root
5. Inside `packages/hash/integration` in another terminal run
  a.  `LOG_LEVEL="debug" yarn jest "ontology" --maxWorkers 1`  
  b.  `LOG_LEVEL="debug" yarn jest "knowledge" --maxWorkers 1`
6. Also try `yarn dev` and check that login works, and that the `/type-editor` route is not broken.
